src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -2618,6 +2618,101 @@ internal static int IndexOfAnyValueType<T>(ref T searchSpace, T value0, T value1
             return -1;
         }
 
+        public static void Replace<T>(Span<T> span, T oldValue, T newValue)
+        {
+            for (int i = 0; i < span.Length; ++i)
+            {
+                ref T val = ref span[i];
+
+                if (EqualityComparer<T>.Default.Equals(oldValue, val))
+                {
+                    val = newValue;
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ReplaceValueType<T>(ref T src, ref T dst, T oldValue, T newValue, nuint length) where T : struct
```

Comment:
For `string.Replace(char, char)` to work, the signature has to look like this. Cf. Cf. https://github.com/dotnet/runtime/issues/75322#issuecomment-1252842175

Codegen-wise (at least on windows) the `length` is passed via stack. This method will be inlined, so no problem and below for the vectorized code-path (not inlined) the cost will be amortized. So I think this isn't a problem.

Reply:
For my edification, how much worse it it if string.Replace is implemented as a memcpy and then an in-place replace?  2x worse? Less than that?  I'm wondering whether the two-arg case is important enough that we should be exposing it publicly, or if the in-place replace is sufficient.  We went with only the two-arg case for endianness reversal.

Reply:
The change would look like 64c7e8136aad79f074842b3ac3fc95aea78a35f8.

For the benchmarks I just tested the worst-case*, i.e. the first char is a match, so the whole string gets copied and then again from the beginning iterated over to replace -- $O(2n)$ instead of $O(n)$ with the out-of-place approach (current PR).

\* string consisting all of `-`, then `string.Replace('-', '+')`

Numbers look like (AVX2 machine):

```
|     Method | Length |      Mean |    Median | Ratio |
|----------- |------- |----------:|----------:|------:|
| OutOfPlace |      7 |  34.65 ns |  34.10 ns |  1.00 |
|    Inplace |      7 |  34.32 ns |  33.12 ns |  1.00 |
|            |        |           |           |       |
| OutOfPlace |      8 |  26.95 ns |  26.08 ns |  1.00 |
|    Inplace |      8 |  27.53 ns |  27.27 ns |  1.01 |
|            |        |           |           |       |
| OutOfPlace |     15 |  24.33 ns |  24.33 ns |  1.00 |
|    Inplace |     15 |  45.03 ns |  43.91 ns |  1.85 |
|            |        |           |           |       |
| OutOfPlace |     16 |  25.90 ns |  24.98 ns |  1.00 |
|    Inplace |     16 |  32.42 ns |  32.27 ns |  1.16 |
|            |        |           |           |       |
| OutOfPlace |     31 |  26.86 ns |  26.69 ns |  1.00 |
|    Inplace |     31 |  36.23 ns |  36.21 ns |  1.35 |
|            |        |           |           |       |
| OutOfPlace |    100 |  37.64 ns |  37.18 ns |  1.00 |
|    Inplace |    100 |  50.73 ns |  50.67 ns |  1.35 |
|            |        |           |           |       |
| OutOfPlace |    500 | 108.17 ns | 107.71 ns |  1.00 |
|    Inplace |    500 | 131.40 ns | 129.90 ns |  1.22 |
|            |        |           |           |       |
| OutOfPlace |   1000 | 192.27 ns | 191.45 ns |  1.00 |
|    Inplace |   1000 | 223.64 ns | 221.66 ns |  1.19 |
```

For the best-case, i.e. no match, the runtime should be equal, as the IndexOf(oldValue)-scan is the only thing executing.
Then of course there are all variations in between.

PS: numbers are a bit flaky on re-runs, but the given numbers are quite representative (no my machine).